### PR TITLE
Improve `rustc_parse::Parser`'s debuggability

### PR DESCRIPTION
**Pull request by workingjubilee**
The main event is the final commit where I add `Parser::debug_lookahead`. Everything else was basically cleaning up things that bugged me (debugging, as it were) until I felt comfortable enough to actually work on it.

The motivation is that it's annoying as hell to try to figure out how the debug infra works in rustc without having basic queries like `debug!(?parser);` come up "empty". However, Parser has a lot of fields that are mostly irrelevant for most debugging, like the entire ParseSess. I think `Parser::debug_lookahead` with a capped lookahead might be fine as a general-purpose Debug impl, but this adapter version was suggested to allow more choice, and admittedly, it's a refined version of what I was already handrolling just to get some insight going.

### CODE REVIEW

File: compiler/rustc_parse/src/parser/mod.rs [hunk 1]
```diff
@@ -211,9 +211,10 @@ pub type ReplaceRange = (Range<u32>, Vec<(FlatToken, Spacing)>);
 /// Controls how we capture tokens. Capturing can be expensive,
 /// so we try to avoid performing capturing in cases where
 /// we will never need an `AttrTokenStream`.
-#[derive(Copy, Clone)]
+#[derive(Copy, Clone, Default)]
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/mod.rs]: Is this default used anywhere? If not, please let's keep this explicit. I think "this is the default mode" doesn't mean "this needs a `Default` impl", it just means "this is what's used in `Parser::new`").

File: compiler/rustc_parse/src/parser/mod.rs [hunk 2]
```diff
@@ -121,7 +121,7 @@ macro_rules! maybe_recover_from_interpolated_ty_qpath {
     };
 }
 
-#[derive(Clone, Copy, Debug)]
+#[derive(Clone, Copy, Debug, PartialEq)]
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/mod.rs:124]: Why? I would prefer if people just write `matches!` here.

File: compiler/rustc_parse/src/parser/mod.rs [hunk 3]
```diff
@@ -1534,6 +1538,47 @@ impl<'a> Parser<'a> {
...
+                        current
+                    });
+                dbg_fmt.field_with("tokens", |field| field.debug_list().entries(tokens).finish());
+                dbg_fmt.field("approx_token_stream_pos", &parser.num_bump_calls);
+
+                // some fields are interesting for certain values, as they relate to macro parsing
+                if let Some(subparser) = parser.subparser_name {
+                    dbg_fmt.field("subparser_name", &subparser);
+                }
+                if parser.recovery == Recovery::Forbidden {
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/mod.rs:1570]: Make this an exhaustive match or a `matches!()`

**Review by compiler-errors — CHANGES_REQUESTED** [review]
I'd prefer if you dropped the changews for `#[derive(Default)]` for `Collect`, and `#[derive(PartialEq)]` for `Recovery`.

File: compiler/rustc_parse/src/parser/mod.rs (hunk 1 above)
Comment (workingjubilee) [reply compiler/rustc_parse/src/parser/mod.rs]: Mmkay!

File: compiler/rustc_parse/src/parser/mod.rs (hunk 2 above)
Comment (workingjubilee) [reply compiler/rustc_parse/src/parser/mod.rs:124]: tbh, no particularly strong reason, my bias is just to prefer to implement traits that seem applicable unless I can articulate my reservations about doing so, and I couldn't figure out that here.

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**